File: desktop/lib/models/sales_return.dart
import 'entities.dart';

/// Where a sales return can be raised from.
///
/// The goods physically left on a delivery note and the money was billed on a
/// sales invoice, so either is a starting point: a customer who sends goods
/// back before being invoiced has only the first.
enum SalesReturnSource {
  deliveryNote('DELIVERY_NOTE', 'Delivery note'),
  salesInvoice('SALES_INVOICE', 'Sales invoice');

  const SalesReturnSource(this.code, this.label);

  final String code;
  final String label;

  static SalesReturnSource fromCode(String code) =>
      SalesReturnSource.values.firstWhere(
        (value) => value.code == code,
        orElse: () => SalesReturnSource.deliveryNote,
      );
}

/// One line of a customer return.
class SalesReturnLine {
  const SalesReturnLine({
    required this.id,
    required this.lineNumber,
    required this.sourceType,
    required this.sourceDocumentNumber,
    required this.sourceDocumentLineNumber,
    required this.productId,
    required this.description,
    required this.dispatchedQuantity,
    required this.alreadyReturnedQuantity,
    required this.currentReturnQuantity,
    required this.restockQuantity,
    required this.damagedQuantity,
    required this.scrapQuantity,
    required this.reasonCode,
    required this.unitPrice,
    required this.taxAmount,
    required this.netAmount,
    required this.batchNumber,
    required this.remarks,
  });

  final String id;
  final int lineNumber;
  final SalesReturnSource sourceType;
  final String sourceDocumentNumber;
  final int sourceDocumentLineNumber;
  final String productId;

  /// What the source document called it. Shown instead of the product id: a
  /// reader wants the goods named, not a UUID.
  final String description;
  final String dispatchedQuantity;
  final String alreadyReturnedQuantity;
  final String currentReturnQuantity;

  /// How much of it can be sold again. The rest came back damaged or as scrap:
  /// still owned and still worth what it cost, but not on the shelf.
  final String restockQuantity;
  final String damagedQuantity;
  final String scrapQuantity;
  final String reasonCode;
  final String unitPrice;
  final String taxAmount;
  final String netAmount;
  final String batchNumber;
  final String remarks;

  /// What is still returnable against the source line after this one.
  String get pendingQuantity {
    final double dispatched = double.tryParse(dispatchedQuantity) ?? 0;
    final double already = double.tryParse(alreadyReturnedQuantity) ?? 0;
    final double current = double.tryParse(currentReturnQuantity) ?? 0;
    final double pending = dispatched - already - current;
    return (pending < 0 ? 0 : pending).toStringAsFixed(4);
  }

  factory SalesReturnLine.fromJson(Json json) => SalesReturnLine(
        id: stringValue(json['id']),
        lineNumber: (json['line_number'] as num?)?.toInt() ?? 0,
        sourceType:
            SalesReturnSource.fromCode(stringValue(json['source_document_type'])),
        sourceDocumentNumber: stringValue(json['source_document_number']),
        sourceDocumentLineNumber:
            (json['source_document_line_number'] as num?)?.toInt() ?? 0,
        productId: stringValue(json['product_id']),
        description: stringValue(json['description']),
        dispatchedQuantity: stringValue(json['dispatched_quantity']),
        alreadyReturnedQuantity: stringValue(json['already_returned_quantity']),
        currentReturnQuantity: stringValue(json['current_return_quantity']),
        restockQuantity: stringValue(json['restock_quantity']),
        damagedQuantity: stringValue(json['damaged_quantity']),
        scrapQuantity: stringValue(json['scrap_quantity']),
        reasonCode: stringValue(json['reason_code']),
        unitPrice: stringValue(json['unit_price']),
        taxAmount: stringValue(json['tax_amount']),
        netAmount: stringValue(json['net_amount']),
        batchNumber: stringValue(json['batch_number']),
        remarks: stringValue(json['remarks']),
      );
}

/// Goods coming back from a customer.
class SalesReturn {
  const SalesReturn({
    required this.id,
    required this.customerId,
    required this.branchId,
    required this.warehouseId,
    required this.returnNumber,
    required this.returnDate,
    required this.customerReturnNumber,
    required this.returnReason,
    required this.status,
    required this.totalCurrentReturnQuantity,
    required this.totalRestockQuantity,
    required this.subtotal,
    required this.taxTotal,
    required this.grandTotal,
    required this.journalEntryId,
    required this.costJournalEntryId,
    required this.cancelReason,
    required this.remarks,
    required this.lines,
  });

  final String id;
  final String customerId;
  final String branchId;
  final String warehouseId;
  final String returnNumber;
  final String returnDate;
  final String customerReturnNumber;
  final String returnReason;
  final String status;
  final String totalCurrentReturnQuantity;
  final String totalRestockQuantity;
  final String subtotal;
  final String taxTotal;

  /// What the customer is credited, tax included.
  final String grandTotal;

  /// The journal that credited the customer, and the one that put the cost of
  /// the goods back into stock. Two, because they answer two questions: the
  /// credit is at the selling price, the stock returns at what it cost.
  final String journalEntryId;
  final String costJournalEntryId;
  final String cancelReason;
  final String remarks;
  final List<SalesReturnLine> lines;

  bool get isDraft => status == 'DRAFT';
  bool get isApproved => status == 'APPROVED';
  bool get isCompleted => status == 'COMPLETED';
  bool get isCancelled => status == 'CANCELLED';
  bool get isClosed => status == 'CLOSED';

  /// Whether the goods are actually back and the customer actually credited.
  /// A draft or an approved return has moved nothing yet.
  bool get hasMoved => isCompleted || isClosed;

  factory SalesReturn.fromJson(Json json) => SalesReturn(
        id: stringValue(json['id']),
        customerId: stringValue(json['customer_id']),
        branchId: stringValue(json['branch_id']),
        warehouseId: stringValue(json['warehouse_id']),
        returnNumber: stringValue(json['return_number']),
        returnDate: stringValue(json['return_date']),
        customerReturnNumber: stringValue(json['customer_return_number']),
        returnReason: stringValue(json['return_reason']),
        status: stringValue(json['status']),
        totalCurrentReturnQuantity:
            stringValue(json['total_current_return_quantity']),
        totalRestockQuantity: stringValue(json['total_restock_quantity']),
        subtotal: stringValue(json['subtotal']),
        taxTotal: stringValue(json['tax_total']),
        grandTotal: stringValue(json['grand_total']),
        journalEntryId: stringValue(json['journal_entry_id']),
        costJournalEntryId: stringValue(json['cost_journal_entry_id']),
        cancelReason: stringValue(json['cancel_reason']),
        remarks: stringValue(json['remarks']),
        lines: [
          for (final dynamic line in json['lines'] is List ? json['lines'] : const [])
            if (line is Map) SalesReturnLine.fromJson(Map<String, dynamic>.from(line)),
        ],
      );
}

/// A document a return can be raised against, flattened for the picker.
///
/// Delivery notes and sales invoices are different resources with different
/// field names; the editor only needs what they have in common plus their
/// returnable lines.
class ReturnableDocument {
  const ReturnableDocument({
    required this.id,
    required this.sourceType,
    required this.number,
    required this.documentDate,
    required this.customerId,
    required this.lines,
  });

  final String id;
  final SalesReturnSource sourceType;
  final String number;
  final String documentDate;
  final String customerId;
  final List<ReturnableLine> lines;

  String get label => '$number  ·  $documentDate';

  /// Read a delivery note, whose dispatched quantity is what went out.
  factory ReturnableDocument.fromDeliveryNote(Json json) => ReturnableDocument(
        id: stringValue(json['id']),
        sourceType: SalesReturnSource.deliveryNote,
        number: stringValue(json['delivery_note_number']),
        documentDate: stringValue(json['delivery_date']),
        customerId: stringValue(json['customer_id']),
        lines: _lines(json, 'current_delivery_quantity'),
      );

  /// Read a sales invoice, whose invoiced quantity is what was billed.
  factory ReturnableDocument.fromSalesInvoice(Json json) => ReturnableDocument(
        id: stringValue(json['id']),
        sourceType: SalesReturnSource.salesInvoice,
        number: stringValue(json['invoice_number']),
        documentDate: stringValue(json['invoice_date']),
        customerId: stringValue(json['customer_id']),
        lines: _lines(json, 'current_invoice_quantity'),
      );

  static List<ReturnableLine> _lines(Json json, String quantityKey) => [
        for (final dynamic line in json['lines'] is List ? json['lines'] : const [])
          if (line is Map)
            ReturnableLine.fromJson(
              Map<String, dynamic>.from(line),
              quantityKey: quantityKey,
            ),
      ];
}

/// One line of a document that can be returned against.
class ReturnableLine {
  const ReturnableLine({
    required this.id,
    required this.lineNumber,
    required this.productId,
    required this.description,
    required this.quantity,
    required this.unitPrice,
  });

  final String id;
  final int lineNumber;
  final String productId;
  final String description;

  /// What this line sent the customer, which is the ceiling on the return.
  final String quantity;
  final String unitPrice;

  String get label {
    final String name = description.isEmpty ? 'Line $lineNumber' : description;
    return '$lineNumber. $name  ·  $quantity';
  }

  factory ReturnableLine.fromJson(Json json, {required String quantityKey}) =>
      ReturnableLine(
        id: stringValue(json['id']),
        lineNumber: (json['line_number'] as num?)?.toInt() ?? 0,
        productId: stringValue(json['product_id']),
        description: stringValue(json['description']),
        quantity: stringValue(json[quantityKey]),
        unitPrice: stringValue(json['unit_price']),
      );
}

File: backend/app/sales_return/schemas/sales_return.py
"""Validated contracts for sales returns."""

from datetime import date, datetime
from decimal import Decimal
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class SalesReturnSchema(BaseModel):
    """Apply strict input and ORM response behavior."""

    model_config = ConfigDict(extra="forbid", from_attributes=True)


class SalesReturnStatus(StrEnum):
    """Supported sales return lifecycle statuses."""

    DRAFT = "DRAFT"
    APPROVED = "APPROVED"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"
    CLOSED = "CLOSED"


class SalesReturnSourceType(StrEnum):
    """Documents a sales return can be raised from.

    The goods physically left on a delivery note and the money was billed on a
    sales invoice, so either is a legitimate starting point: a customer who
    returns goods before being invoiced has a delivery note and nothing else.
    """

    DELIVERY_NOTE = "DELIVERY_NOTE"
    SALES_INVOICE = "SALES_INVOICE"


class SalesReturnAttachmentWrite(SalesReturnSchema):
    """Carry one sales return attachment into a request."""

    file_name: str = Field(min_length=1, max_length=260)
    mime_type: str | None = Field(default=None, max_length=120)
    file_path: str = Field(min_length=1, max_length=1024)
    attachment_kind: str = Field(
        default="SALES_RETURN_FILE", min_length=1, max_length=40
    )


class SalesReturnNoteWrite(SalesReturnSchema):
    """Carry one sales return note into a request."""

    note_type: str = Field(default="INTERNAL", min_length=1, max_length=30)
    note: str = Field(min_length=1)


class SalesReturnSourceWrite(SalesReturnSchema):
    """Carry one sales return source into a request."""

    source_document_type: SalesReturnSourceType
    source_document_id: UUID


class SalesReturnLineWrite(SalesReturnSchema):
    """Carry one sales return line into a request."""

    source_document_type: SalesReturnSourceType
    source_document_id: UUID
    source_document_line_id: UUID
    line_number: int = Field(ge=1)
    current_return_quantity: Decimal = Field(gt=0, max_digits=18, decimal_places=4)
    #: How much of it can be sold again. Defaults to all of it on the reading
    #: that goods come back fit unless somebody says otherwise, which is what a
    #: warehouse clerk booking a return in a hurry means.
    restock_quantity: Decimal | None = Field(
        default=None, ge=0, max_digits=18, decimal_places=4
    )
    damaged_quantity: Decimal = Field(
        default=Decimal("0"), ge=0, max_digits=18, decimal_places=4
    )
    scrap_quantity: Decimal = Field(
        default=Decimal("0"), ge=0, max_digits=18, decimal_places=4
    )
    reason_code: str | None = Field(default=None, max_length=80)
    item_condition: str | None = Field(default=None, max_length=80)
    is_damaged: bool = False
    is_expired: bool = False
    #: Left unset, the line is credited at what the source document charged.
    #: Sent explicitly -- zero included -- that is what the customer gets: a
    #: free replacement is credited at nothing, not at the price of the goods
    #: it replaced.
    unit_price: Decimal | None = Field(
        default=None, ge=0, max_digits=18, decimal_places=4
    )
    #: None means the caller said nothing, so the customer's standing
    #: discount applies. Zero means they said no discount.
    discount_percent: Decimal | None = Field(
        default=None, ge=0, le=100, max_digits=9, decimal_places=4
    )
    discount_amount: Decimal | None = Field(
        default=None, ge=0, max_digits=18, decimal_places=4
    )
    charges_amount: Decimal = Field(
        default=Decimal("0"), ge=0, max_digits=18, decimal_places=4
    )
    tax_profile_id: UUID | None = None
    packaging_type_id: UUID | None = None
    sales_uom_id: UUID | None = None
    return_uom_id: UUID | None = None
    conversion_factor: Decimal | None = Field(
        default=None, gt=0, max_digits=24, decimal_places=10
    )
    warehouse_id: UUID | None = None
    storage_node_id: UUID | None = None
    batch_number: str | None = Field(default=None, max_length=120)
    expiry_date: date | None = None
    manufacturing_date: date | None = None
    remarks: str | None = None

    @model_validator(mode="after")
    def _condition_adds_up(self) -> "SalesReturnLineWrite":
        """Refuse a line that says more came back than came back.

        The three buckets decide where the goods land and what the firm can
        sell, so a line whose parts exceed its whole would put stock on the
        shelf that never arrived.
        """
        restock = (
            self.current_return_quantity - self.damaged_quantity - self.scrap_quantity
            if self.restock_quantity is None
            else self.restock_quantity
        )
        if restock < 0:
            raise ValueError(
                "Damaged and scrap quantities cannot exceed the returned quantity."
            )
        total = restock + self.damaged_quantity + self.scrap_quantity
        if total != self.current_return_quantity:
            raise ValueError(
                "Restock, damaged and scrap quantities must add up to the "
                f"returned quantity ({self.current_return_quantity})."
            )
        return self


class SalesReturnCreate(SalesReturnSchema):
    """Create one sales return."""

    customer_id: UUID | None = None
    branch_id: UUID | None = None
    business_profile_id: UUID | None = None
    warehouse_id: UUID
    return_date: date
    customer_return_number: str | None = Field(default=None, max_length=120)
    customer_return_date: date | None = None
    reference_delivery_note_number: str | None = Field(default=None, max_length=80)
    reference_invoice_number: str | None = Field(default=None, max_length=80)
    return_reason: str | None = Field(default=None, max_length=80)
    currency_code: str | None = Field(default=None, max_length=10)
    exchange_rate: Decimal | None = Field(
        default=None, gt=0, max_digits=18, decimal_places=6
    )
    reference_number: str | None = Field(default=None, max_length=120)
    remarks: str | None = None
    allow_over_return: bool = False
    over_return_percent: Decimal = Field(
        default=Decimal("0"), ge=0, max_digits=9, decimal_places=4
    )
    additional_charges: Decimal = Field(
        default=Decimal("0"), ge=0, max_digits=18, decimal_places=4
    )
    round_off: Decimal = Field(default=Decimal("0"), max_digits=18, decimal_places=4)
    return_number: str | None = Field(default=None, max_length=60)
    source_documents: list[SalesReturnSourceWrite] = Field(
        default_factory=list, max_length=100
    )
    lines: list[SalesReturnLineWrite] = Field(min_length=1, max_length=1000)
    attachments: list[SalesReturnAttachmentWrite] = Field(
        default_factory=list, max_length=500
    )
    notes: list[SalesReturnNoteWrite] = Field(default_factory=list, max_length=500)

    @field_validator("return_number", mode="before")
    @classmethod
    def _normalize_number(cls, value: str | None) -> str | None:
        if value is None:
            return None
        token = value.strip().upper()
        return token or None


class SalesReturnUpdate(SalesReturnCreate):
    """Replace one sales return."""


class SalesReturnImportRequest(SalesReturnSchema):
    """Import a validated batch of sales returns.

    Bounded at 500 for the same reason the purchase-return batch is: every
    record resolves its source delivery note or invoice, prices its lines and
    simulates tax per line, so a batch is not a cheap loop over inserts.
    """

    records: list[SalesReturnCreate] = Field(min_length=1, max_length=500)


class SalesReturnListFilters(SalesReturnSchema):
    """Narrow a sales return list."""

    customer_id: UUID | None = None
    branch_id: UUID | None = None
    warehouse_id: UUID | None = None
    status: SalesReturnStatus | None = None
    return_from: date | None = None
    return_to: date | None = None
    include_deleted: bool = False

    @model_validator(mode="after")
    def _dates_are_in_order(self) -> "SalesReturnListFilters":
        """Refuse a window that ends before it starts."""
        if (
            self.return_from is not None
            and self.return_to is not None
            and self.return_to < self.return_from
        ):
            raise ValueError("return_to cannot be earlier than return_from.")
        return self


class SalesReturnAttachmentResponse(SalesReturnSchema):
    """Return one sales return attachment."""

    id: UUID
    sales_return_id: UUID
    file_name: str
    mime_type: str | None
    file_path: str
    attachment_kind: str
    created_at: datetime
    updated_at: datetime


class SalesReturnNoteResponse(SalesReturnSchema):
    """Return one sales return note."""

    id: UUID
    note_type: str
    note: str
    created_at: datetime
    updated_at: datetime


class SalesReturnSourceResponse(SalesReturnSchema):
    """Return one sales return source."""

    id: UUID
    source_document_type: SalesReturnSourceType
    source_document_id: UUID
    source_document_number: str
    source_document_date: date
    customer_id: UUID
    branch_id: UUID
    created_at: datetime
    updated_at: datetime


class SalesReturnLineResponse(SalesReturnSchema):
    """Return one sales return line."""

    id: UUID
    sales_return_id: UUID
    line_number: int
    source_document_type: SalesReturnSourceType
    source_document_id: UUID
    source_document_number: str
    source_document_line_id: UUID
    source_document_line_number: int
    product_id: UUID
    description: str | None
    dispatched_quantity: Decimal
    already_returned_quantity: Decimal
    current_return_quantity: Decimal
    restock_quantity: Decimal
    damaged_quantity: Decimal
    scrap_quantity: Decimal
    reason_code: str | None
    item_condition: str | None
    is_damaged: bool
    is_expired: bool
    unit_price: Decimal
    discount_percent: Decimal
    discount_amount: Decimal
    charges_amount: Decimal
    gross_amount: Decimal
    tax_profile_id: UUID | None
    tax_amount: Decimal
    net_amount: Decimal
    packaging_type_id: UUID | None
    sales_uom_id: UUID | None
    return_uom_id: UUID | None
    conversion_factor: Decimal
    conversion_version: int | None
    warehouse_id: UUID | None
    storage_node_id: UUID | None
    batch_number: str | None
    batch_id: UUID | None
    expiry_date: date | None
    manufacturing_date: date | None
    inventory_transaction_id: UUID | None
    remarks: str | None
    created_at: datetime
    updated_at: datetime


class SalesReturnResponse(SalesReturnSchema):
    """Return one sales return."""

    id: UUID
    firm_id: UUID
    customer_id: UUID
    branch_id: UUID
    warehouse_id: UUID
    salesman_id: UUID | None
    territory_id: UUID | None
    business_profile_id: UUID | None
    return_number: str
    return_date: date
    customer_return_number: str | None
    customer_return_date: date | None
    reference_delivery_note_number: str | None
    reference_invoice_number: str | None
    return_reason: str | None
    currency_code: str | None
    exchange_rate: Decimal | None
    reference_number: str | None
    remarks: str | None
    allow_over_return: bool
    over_return_percent: Decimal
    status: SalesReturnStatus
    total_source_quantity: Decimal
    total_already_returned_quantity: Decimal
    total_current_return_quantity: Decimal
    total_restock_quantity: Decimal
    line_discount_total: Decimal
    subtotal: Decimal
    tax_total: Decimal
    additional_charges: Decimal
    round_off: Decimal
    grand_total: Decimal
    journal_entry_id: UUID | None
    cost_journal_entry_id: UUID | None
    approved_at: datetime | None
    completed_at: datetime | None
    closed_at: datetime | None
    cancel_reason: str | None
    close_reason: str | None
    is_deleted: bool
    created_at: datetime
    updated_at: datetime
    version: int
    lines: list[SalesReturnLineResponse]
    sources: list[SalesReturnSourceResponse]
    attachments: list[SalesReturnAttachmentResponse]
    notes: list[SalesReturnNoteResponse]


class SalesReturnSummary(SalesReturnSchema):
    """Summarise sales returns for one firm."""

    total_returns: int
    draft_returns: int
    approved_returns: int
    completed_returns: int
    cancelled_returns: int
    total_return_value: Decimal
    total_restock_quantity: Decimal


class SalesReturnRegisterRecord(SalesReturnSchema):
    """One row of the sales return register report."""

    return_id: UUID
    return_number: str
    customer_return_number: str | None
    customer_id: UUID
    branch_id: UUID
    warehouse_id: UUID
    return_date: date
    grand_total: Decimal
    status: SalesReturnStatus


class SalesReturnByCustomerRecord(SalesReturnSchema):
    """Returned value and count per customer."""

    customer_id: UUID
    customer_name: str
    return_amount: Decimal
    return_count: int


class SalesReturnByProductRecord(SalesReturnSchema):
    """Returned quantity and value per product."""

    product_id: UUID
    product_code: str
    product_name: str
    return_quantity: Decimal
    restock_quantity: Decimal
    return_amount: Decimal
    return_count: int


class SalesReturnReconciliationRecord(SalesReturnSchema):
    """One return line set against the document it was dispatched on."""

    return_id: UUID
    return_number: str
    return_date: date
    source_document_type: SalesReturnSourceType
    source_document_id: UUID
    source_document_number: str
    source_document_line_id: UUID
    source_document_line_number: int
    product_id: UUID
    product_name: str
    dispatched_quantity: Decimal
    already_returned_quantity: Decimal
    current_return_quantity: Decimal
    pending_quantity: Decimal
    restock_quantity: Decimal
    reason_code: str | None
    is_damaged: bool
    is_expired: bool
